### ISSUE TITLE
Improved results with no SCINet module

### ISSUE THREAD

**Issue by mcvageesh** [cure-lab/SCINet#14]
Hello authors / fellow GitHub members,

In models/SCINet.py, in class SCINet, forward() method, comment the lines 334, 335, 336 - we  do not pass the input x through the self.blocks1 and the subsequent residual addition step,  i.e. we only have x = self.projection1(x) in the first stack. Also, use stacks=1 (there is no need for two stacks). Now, run the experiments on the datasets.

I believe this will outperform the SCINet model. 

Do try it out and let me know - it seems the odd-even splitting and interactive learning only worsen the results (as compared to what I have suggested, which, at its core, is a simple linear model).

**Comment by VEWOXIC** [comment]
We are very glad to receive your feedback. 
But as you mentioned, after the modification, the whole network is basically a simple fully connected layer. If this works better than the SOTA model, it will make a profound impact on the research community! 
We are very excited about your finding. For our further investigation, can you please tell us your training configuration? E.g. the dataset, the configuration of the model (or the command line you just executed). Also, it would be great if you can share your final result with us!

**Comment by mcvageesh** [comment]
I have tested on the ETT datasets. The commands are the same as you describe in the readme , except --stacks=1 always for me.  For your reference, I have also attached my SCINet.py file. I will also try to upload my results in a while.

Also, I would argue, the whole network is not a simple fully connected layer at the end - 
self.projection1 = nn.Conv1d(self.input_len, self.output_len, kernel_size=1, stride=1, bias=False)
The in_channel argument is self.input_len and kernel_size is 1.  What the conv1d layer is doing is not exactly the same as a dense layer would do - i.e. if instead, we 
1) flatten out the input x
2) use nn.Linear(self.input_len*self.input_dim, self.output_len*output_dim) on x
3) reshape the  output to (self.output_len, output_dim)

That would be a true dense layer, wouldn't it? Do let me know what you think.

[SCINet modified.txt](https://github.com/cure-lab/SCINet/files/7631828/SCINet.modified.txt)

**Comment by mcvageesh** [comment]
Here are the results for multivariate testing on all the ETT datasets (see attached image). I do not have enough computational power and time to run the PEMS and financial datasets, so please check the performance there too - I believe it will show an improvement.
![Results](https://user-images.githubusercontent.com/54277269/144231224-097b4351-85cf-4dd6-b4e3-a37e60ecf2b9.PNG)

**Comment by VEWOXIC** [comment]
Sorry for the delay. 
We have tested this modified model on both ETT and PeMS dataset. On the multivariant task of ETT, the modified version outperforms our model on almost all the horizon lengths. But it only performs better on horizon lengths of 24, and 48 in Univariant task. It's still very impressive considering it's only a weighted sum of the input. But as shown in the chart, on PeMS dataset, the outcome is not very ideal. 
![image](https://user-images.githubusercontent.com/49720945/144551725-f3f55412-b1a8-4132-94af-0f8399989e81.png)
![image](https://user-images.githubusercontent.com/49720945/144551735-dbac0481-fb18-4b31-b894-8468a9148f4e.png)

After analyzing the code, we conclude that the single conv1d layer is basically performing a weighted sum on the time dimension of input values. We also visualized the weight of this layer and found the modified network is served as a Naive predict algorithm that heavily relies on the last time step of the input and some seasonality information. 
Following is the weight heatmap of the 'Univariant, input 64, output 24' setting. The y axis stands for output set and x axis stands for input steps. E.g. the first line tells us how the input steps are combined to get the output step 0. As the weights tell us, this model mainly focuses on the last input step. 
![image](https://user-images.githubusercontent.com/49720945/144552364-b12f3026-10a1-44c4-b91f-e5f44fe382f3.png)
Following is the weight heatmap of the 'Multivariant, input 48, output 24' setting. There's a very obvious seasonal pattern in this heatmap which the period is exactly 24. Considering this dataset is the data acquired every hour in a period, we assume that this model extracted the daily pattern in it! 
![image](https://user-images.githubusercontent.com/49720945/144553009-28d774c7-9831-422c-b515-323ad9adf06c.png)

By inspecting the weight heatmap of PEMS, we found there's no pattern for the model to learn. Thus, it just give a large weight to the last input step. 
![image](https://user-images.githubusercontent.com/49720945/144553959-937b0757-b5a1-4850-a8a3-df66334215c5.png)
![image](https://user-images.githubusercontent.com/49720945/144553978-dac5c0d2-97a5-49b0-8c51-dc76812c5154.png)
![image](https://user-images.githubusercontent.com/49720945/144553995-1489cd89-a756-43e3-9905-05af4b70d88b.png)
![image](https://user-images.githubusercontent.com/49720945/144554014-628dbc43-1968-40ed-a944-e132aa407bd5.png)

Anyway, your finding provided us with a very significant insight into this area! We plan to explore further on this topic! Many Thanks for your finding!!!

**Comment by mcvageesh** [comment]
Hi,
Thank you for the detailed analysis. Yes, it seems that simpler methods will outperform deep models in many time series forecasting applications. 

For the PEMS (and probably financial) datasets, I think there are too many time series for the simple model to perform well. If you notice, the conv1D layer also adds a constraint that a single filter must be able to capture the information from all the time series to predict a single future step for each of the time series. 

Maybe :
1) flatten out the input x
2) use nn.Linear(self.input_len * self.input_dim, self.output_len * output_dim) on x
3) reshape the output to (self.output_len, output_dim)

i.e. a fully connected dense layer could be used as a benchmark for these datasets.

For the ETT datasets, as the benchmark, you can use what I suggested earlier, but be sure to also set the bias=True in the conv1D layer. You will notice a significant improvement in the univariate forecasting case with large horizon lengths.

I would be interested to read your results and analysis in the updated paper. Good luck :) I will close this issue now.

**Comment by d-zh** [comment]
Hi,
   Nice analysis! I wonder how to get these heatmaps, could you provide some codes?

**Comment by Lonearly** [comment]
> 我们已经在ETT和PeMS数据集上测试了这个修改后的模型。在ETT的多变量任务中，修改后的版本在几乎所有的地平线长度上都优于我们的模型。但它仅在 24 的地平线长度上表现更好，在单变量任务中为 48。考虑到它只是输入的加权总和，它仍然非常令人印象深刻。但如图所示，在PeMS数据集上，结果不是很理想。 ![image](https://user-images.githubusercontent.com/49720945/144551725-f3f55412-b1a8-4132-94af-0f8399989e81.png) ![image](https://user-images.githubusercontent.com/49720945/144551735-dbac0481-fb18-4b31-b894-8468a9148f4e.png)
> 
> 在分析了代码之后，我们得出结论，单个conv1d层基本上是在输入值的时间维度上执行加权和。我们还可视化了该层的权重，发现修改后的网络被用作朴素预测算法，该算法严重依赖于输入的最后一个时间步长和一些季节性信息。以下是“Univariant，输入 64，输出 24”设置的权重热图。y 轴代表输出集，x 轴代表输入步进。例如，第一行告诉我们如何组合输入步骤以获得输出步骤 0。正如权重告诉我们的那样，该模型主要关注最后一个输入步骤。 ![image](https://user-images.githubusercontent.com/49720945/144552364-b12f3026-10a1-44c4-b91f-e5f44fe382f3.png)以下是“多变量，输入 48，输出 24”设置的权重热图。在这个热图中有一个非常明显的季节性模式，周期正好是 24。考虑到这个数据集是一段时间内每小时获取的数据，我们假设这个模型提取了其中的每日模式！ ![image](https://user-images.githubusercontent.com/49720945/144553009-28d774c7-9831-422c-b515-323ad9adf06c.png)
> 
> 通过检查PEMS的权重热图，我们发现模型没有可以学习的模式。因此，它只是给最后一个输入步骤很大的权重。 ![image](https://user-images.githubusercontent.com/49720945/144553959-937b0757-b5a1-4850-a8a3-df66334215c5.png) ![image](https://user-images.githubusercontent.com/49720945/144553978-dac5c0d2-97a5-49b0-8c51-dc76812c5154.png) ![image](https://user-images.githubusercontent.com/49720945/144553995-1489cd89-a756-43e3-9905-05af4b70d88b.png) ![image](https://user-images.githubusercontent.com/49720945/144554014-628dbc43-1968-40ed-a944-e132aa407bd5.png)
> 
> 再次感谢您让我们知道您的发现。我们从中学到的是，尽管数据集的行为看似复杂，但传统方法在许多数据集上的表现都优于深度学习模型。在进入深度学习模型之前，我们应该先尝试这些传统方法（例如朴素预测、ARIMA 等）。我们将在下一版本中相应地更新该论文。
您好，感谢您对此做出的分析，请问该如何绘制这些图呢，十分感谢您的回复！
